### PR TITLE
Fix false start

### PR DESCRIPTION
**Pull request by grahn**
This fixes an issue where messages can be dropped when there are spurious start bytes in the byte stream. See the commit messages and new comments for more detail.

@brandon-hass and @calrsom Let me know what you think.

### CODE REVIEW

File: mavlink/message.go [hunk 1]
```diff
@@ -102,53 +103,96 @@ func newPacketFromBytes(b []byte) (*Packet, int) {
...
+		p.Payload = dec.buffer[payloadStart : payloadStart+payloadLen]
+		crc.Write(p.Payload)
 
-	// does the transmitted checksum match our computed checksum?
-	if p.Checksum != crc.Sum16() {
-		return p, ErrCrcFail
+		crcx, err := dec.Dialects.findCrcX(p.MsgID)
+		if err != nil {
+			dec.buffer = dec.buffer[1:]
+			startFound = false
```
Comment (brandon-hass) [inline mavlink/message.go]: Is it necessary to set `startFound = false` here?

File: mavlink/message.go [hunk 2]
```diff
@@ -102,53 +103,96 @@ func newPacketFromBytes(b []byte) (*Packet, int) {
...
+
+		// does the transmitted checksum match our computed checksum?
+		if p.Checksum != crc.Sum16() {
+			// strip off start byte
+			dec.buffer = dec.buffer[1:]
+			startFound = false
+		} else {
+			dec.CurrSeqID = p.SeqID
+			dec.buffer = dec.buffer[packetLen:]
+			startFound = false
```
Comment (brandon-hass) [inline mavlink/message.go]: Same question as above.

File: mavlink/message.go (hunk 1 above)
Comment (grahn) [reply mavlink/message.go]: It's necessary so that bytes get read from `Decoder.br` if no start byte is found in the buffer. However, `startFound = false` could just be placed at the top of the loop.

Comment (brandon-hass) [reply mavlink/message.go]: That made sense for the non-error cases. I guess I was curious in the error cases specifically since `startFound` is local and it just returns right after that. Your changes work for that, too.

Comment (grahn) [reply mavlink/message.go]: The idea I had is that the caller could check the error and decide whether or not to call `Decode()` again. And `Decode()` would just continue from where it was in the buffer.

Basically if the caller sees a single `ErrUnknownMsgID` and then gets a valid packet on the next call to `Decode()` that would indicate there was possibly a problem in the stream. However, if the caller gets `ErrUnknownMsgID` on many consecutive calls, that could be a sign that the wrong dialect is being used.

Comment (brandon-hass) [reply mavlink/message.go]: I think that makes sense, too. The behavior wouldn't actually change. I was just thinking that the first thing that happens any time `Decode()` is called is that `startFound` gets set to false because of line 106.

**Review by brandon-hass — APPROVED** [review]

File: mavlink/message.go [hunk 3]
```diff
@@ -102,53 +103,91 @@ func newPacketFromBytes(b []byte) (*Packet, int) {
...
+
+		p.Checksum = bytesToU16(dec.buffer[payloadStart+payloadLen : payloadStart+payloadLen+numChecksumBytes])
+
+		// does the transmitted checksum match our computed checksum?
+		if p.Checksum != crc.Sum16() {
+			// strip off start byte
+			dec.buffer = dec.buffer[1:]
+		} else {
+			dec.CurrSeqID = p.SeqID
+			dec.buffer = dec.buffer[packetLen:]
```
Comment (calrsom) [inline mavlink/message.go:187]: Does this copy bytes back to the front of the slice, or does this continue to grow the array backing the slice?

File: mavlink/message.go [hunk 4]
```diff
@@ -102,53 +103,91 @@ func newPacketFromBytes(b []byte) (*Packet, int) {
...
-	}
 
-	// hdr contains LENGTH, SEQ, SYSID, COMPID, MSGID
-	hdr := make([]byte, 5)
-	if _, err := io.ReadFull(dec.br, hdr); err != nil {
-		return nil, err
-	}
+		if len(dec.buffer) < 2 {
+			// read length byte
+			bytesRead := make([]byte, 1)
```
Comment (calrsom) [inline mavlink/message.go:133]: Possible to read directly into the buffer without allocating another slice?  Not sure how much of a performance impact this makes, maybe it's not a big deal.

File: mavlink/message.go [hunk 5]
```diff
@@ -102,53 +103,91 @@ func newPacketFromBytes(b []byte) (*Packet, int) {
...
+		payloadLen := int(dec.buffer[1])
+		packetLen := hdrLen + payloadLen + numChecksumBytes
+		bytesNeeded := packetLen - len(dec.buffer)
+		if bytesNeeded > 0 {
+			bytesRead := make([]byte, bytesNeeded)
+			n, err := io.ReadAtLeast(dec.br, bytesRead, bytesNeeded)
+			if err != nil {
+				return nil, err
+			}
+			dec.buffer = append(dec.buffer, bytesRead[:n]...)
```
Comment (calrsom) [inline mavlink/message.go:152]: May make sense to extract this pattern into a `appendAtLeast` function; occurs twice here.

File: mavlink/message.go (hunk 3 above)
Comment (grahn) [reply mavlink/message.go:187]: This does neither.

A slice's internal has three elements, a pointer to the backing array, an int for the length, and an int for the capacity. This operation here would just update where the pointer points in the backing array, and update length and capacity accordingly. This would not cause any copying or allocation.

File: mavlink/message.go (hunk 4 above)
Comment (grahn) [reply mavlink/message.go:133]: I ran the following on an edison:

```
func testSliceAlloc() {
    oneByteTime := int64(0)
    multipleByteTime := int64(0)
    numIterations := int64(1000000)

    for i := int64(0); i < numIterations; i++ {
        t := time.Now().UnixNano()
        x := make([]byte, 1)
        oneByteTime += time.Now().UnixNano() - t
        x[0] = 0 // use variable to appease compiler

        t = time.Now().UnixNano()
        x = make([]byte, 263)
        multipleByteTime += time.Now().UnixNano() - t
        x[0] = 0
    }

    fmt.Printf("%d average time ns for make([]byte, 1) with %d iterations\n", oneByteTime/numIterations, numIterations)
    fmt.Printf("%d average time ns for make([]byte, 263) with %d iterations\n", multipleByteTime/numIterations, numIterations)
}
```

and got the following output:

```
2050 average time ns for make([]byte, 1) with 1000000 iterations
2596 average time ns for make([]byte, 263) with 1000000 iterations
```

Let's just round up and say these allocations take 3μs on average.

If the autopilot is connected via RS-232 serial using 115200 for the baud rate, and no parity bit, the time to transfer 263 bytes will be at least `263 * 10 / 115200` which is about 22.8ms, or over 7500 times slower. (263 bytes is used as an example because that's the largest size a mavlink packet can be).

In other words, the memory allocation here is orders of magnitude faster than the receiving of bytes from the autopilot.

Running the same code above using `runtime/pprof` to get a memory profile, only 257 Mallocs are made, i.e., the runtime for the most part is reusing memory it has already allocated.

Anyway, it'd probably be best to use a single circular buffer that only gets allocated once, but in this case, I don't think it matters.

File: mavlink/message.go (hunk 3 above)
Comment (calrsom) [reply mavlink/message.go:187]: I guess maybe I didn't ask the right question.  I was worried there was a memory leak from leaving contents behind in the slice, but the `append` function cleans that up.

File: mavlink/message.go (hunk 4 above)
Comment (calrsom) [reply mavlink/message.go:133]: Cool. Thanks for looking into that. Looks like this is fine then.
